In [1]:
!apt-get install unrar

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
unrar is already the newest version (1:6.1.5-1ubuntu0.1).
0 upgraded, 0 newly installed, 0 to remove and 41 not upgraded.


In [2]:
import os

source_rar_file = "/content/drive/MyDrive/Percobaan_Trainning_Warna/Perpaduan_Warna_Campuran.rar"
destination_directory = "/content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_RAW"

# Create the destination directory if it doesn't exist
os.makedirs(destination_directory, exist_ok=True)

# Use unrar command to extract the contents
!unrar x "{source_rar_file}" "{destination_directory}"

Streaming output truncated to the last 5000 lines.
Extracting  /content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_RAW/Perpaduan_Warna_Campuran/frame_00947.png      27%  OK 
Extracting  /content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_RAW/Perpaduan_Warna_Campuran/frame_00947.txt      27%  OK 
Extracting  /content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_RAW/Perpaduan_Warna_Campuran/frame_00948.png      27%  OK 
Extracting  /content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_RAW/Perpaduan_Warna_Campuran/frame_00948.txt      27%  OK 
Extracting  /content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_RAW/Perpaduan_Warna_Campuran/frame_00949.png      27%  OK 
Extracting  /content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_RAW/Perpaduan_Warna_Campuran/frame_00949.txt      27%  OK 
Extracting  /content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_RAW/Perpaduan_Warna_Campuran/frame_00950.png      27%  OK 
Extracting 

# Task
In order to explore the dataset structure, list the files in the directory `/content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_RAW/Perpaduan_Warna_Campuran` to understand the current organization of image and annotation files.

## Explore Dataset Structure

### Subtask:
Examine the contents of the `/content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_RAW/Perpaduan_Warna_Campuran` directory to understand the current organization of image and annotation files.


**Reasoning**:
To examine the contents of the specified directory, I will first define the directory path, then list its contents using `os.listdir()`, and finally print a sample of the listed files to understand the data organization.



In [3]:
import os

dataset_directory = "/content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_RAW/Perpaduan_Warna_Campuran"

# Get a list of all files and subdirectories within the specified dataset directory
if os.path.exists(dataset_directory):
    files_in_directory = os.listdir(dataset_directory)

    # Print a sample of the files to observe naming conventions and organization
    print(f"Total files found: {len(files_in_directory)}")
    print("First 10 files in the directory:")
    for i, file_name in enumerate(files_in_directory):
        if i >= 10:
            break
        print(file_name)
else:
    print(f"The directory '{dataset_directory}' does not exist.")

Total files found: 6893
First 10 files in the directory:
classes.txt
frame_00000.png
frame_00000.txt
frame_00001.png
frame_00001.txt
frame_00002.png
frame_00002.txt
frame_00003.png
frame_00003.txt
frame_00004.png


## Split Data into Train, Val, Test

### Subtask:
Randomly split the image and annotation files into training, validation, and testing sets and move them into corresponding subdirectories.


**Reasoning**:
First, I will define the input and output directories and the split ratios for the dataset. Then, I will initialize lists to store image and annotation file paths.



In [4]:
import os
import random
import shutil

# 1. Define the input directory containing the image and annotation files
input_data_dir = "/content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_RAW/Perpaduan_Warna_Campuran"

# 2. Define the output base directory where the split data will be stored
output_base_dir = "/content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_SPLIT"

# 3. Define the desired split ratios for training, validation, and testing sets
train_ratio = 0.7
val_ratio = 0.2
test_ratio = 0.1

# Ensure the ratios sum to 1
if not (train_ratio + val_ratio + test_ratio == 1.0):
    raise ValueError("Split ratios must sum to 1.0")

# 4. Initialize empty lists to store the paths of image and annotation files
image_files = []
annotation_files = []


ValueError: Split ratios must sum to 1.0

**Reasoning**:
The previous code failed due to a floating-point precision issue when checking if the split ratios sum to 1.0. I will fix this by calculating the `test_ratio` dynamically from the remaining percentage to ensure the sum is exactly 1.0.



In [5]:
import os
import random
import shutil

# 1. Define the input directory containing the image and annotation files
input_data_dir = "/content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_RAW/Perpaduan_Warna_Campuran"

# 2. Define the output base directory where the split data will be stored
output_base_dir = "/content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_SPLIT"

# 3. Define the desired split ratios for training, validation, and testing sets
train_ratio = 0.7
val_ratio = 0.2
test_ratio = 1.0 - (train_ratio + val_ratio) # Calculate test_ratio to ensure sum is exactly 1.0

# Ensure the ratios sum to 1 (this check is now primarily for verification if individual ratios were set manually)
if not (abs(train_ratio + val_ratio + test_ratio - 1.0) < 1e-9):
    raise ValueError("Split ratios must sum to 1.0")

# 4. Initialize empty lists to store the paths of image and annotation files
image_files = []
annotation_files = []


**Reasoning**:
I will iterate through the input directory to identify and group corresponding image and annotation file pairs while excluding 'classes.txt'.



In [6]:
import os

# 5. Iterate through the input directory to identify and group corresponding image and annotation file pairs
#    Exclude 'classes.txt'.

# Dictionary to store base filenames (e.g., 'frame_00000') and their corresponding image and annotation paths
file_pairs = {}

# List all files in the input directory
all_files = os.listdir(input_data_dir)

for file_name in all_files:
    if file_name == "classes.txt":
        continue # Skip classes.txt

    # Get the base name without extension (e.g., 'frame_00000' from 'frame_00000.png' or 'frame_00000.txt')
    base_name, ext = os.path.splitext(file_name)

    if base_name not in file_pairs:
        file_pairs[base_name] = {"image": None, "annotation": None}

    full_path = os.path.join(input_data_dir, file_name)

    if ext.lower() == '.png': # Assuming images are PNG files
        file_pairs[base_name]["image"] = full_path
    elif ext.lower() == '.txt': # Assuming annotations are TXT files
        file_pairs[base_name]["annotation"] = full_path

# Filter out any base names that don't have both an image and an annotation file
# And convert the dictionary values to a list of (image_path, annotation_path) tuples
valid_file_pairs = [
    (info["image"], info["annotation"])
    for base_name, info in file_pairs.items()
    if info["image"] is not None and info["annotation"] is not None
]

print(f"Found {len(valid_file_pairs)} valid image-annotation pairs.")
print("First 5 valid file pairs:")
for i, pair in enumerate(valid_file_pairs):
    if i >= 5: break
    print(f"  Image: {os.path.basename(pair[0])}, Annotation: {os.path.basename(pair[1])}")

# Populate image_files and annotation_files lists for shuffling later
# This step might be redundant if we directly split from valid_file_pairs, but keeping it for consistency with instructions
for img_path, ann_path in valid_file_pairs:
    image_files.append(img_path)
    annotation_files.append(ann_path)


Found 3445 valid image-annotation pairs.
First 5 valid file pairs:
  Image: frame_00000.png, Annotation: frame_00000.txt
  Image: frame_00001.png, Annotation: frame_00001.txt
  Image: frame_00002.png, Annotation: frame_00002.txt
  Image: frame_00003.png, Annotation: frame_00003.txt
  Image: frame_00004.png, Annotation: frame_00004.txt


**Reasoning**:
Now that the valid image-annotation pairs are identified, I will randomly shuffle this list to ensure an unbiased distribution when splitting the dataset. Then, I will calculate the number of samples for each split (training, validation, and testing) based on the defined ratios.



In [7]:
import random

# 6. Randomly shuffle the list of grouped file pairs
random.shuffle(valid_file_pairs)

# 7. Calculate the number of samples for each split based on the defined ratios
total_samples = len(valid_file_pairs)
train_count = int(total_samples * train_ratio)
val_count = int(total_samples * val_ratio)
test_count = total_samples - train_count - val_count # Ensure test_count covers any remainder due to int conversion

print(f"Total samples: {total_samples}")
print(f"Train samples: {train_count}")
print(f"Validation samples: {val_count}")
print(f"Test samples: {test_count}")

# Split the shuffled list into train, val, and test sets
train_pairs = valid_file_pairs[:train_count]
val_pairs = valid_file_pairs[train_count : train_count + val_count]
test_pairs = valid_file_pairs[train_count + val_count :]


Total samples: 3445
Train samples: 2411
Validation samples: 689
Test samples: 345


**Reasoning**:
I will create the necessary output subdirectories for each split (train, validation, and test) including separate folders for 'images' and 'labels' within each split directory. This sets up the directory structure where the files will be moved.



In [8]:
import os

# 8. Create the necessary output subdirectories
def create_directories(base_dir, splits=["train", "val", "test"]):
    for split in splits:
        os.makedirs(os.path.join(base_dir, split, "images"), exist_ok=True)
        os.makedirs(os.path.join(base_dir, split, "labels"), exist_ok=True)
    print(f"Created directory structure under {base_dir}")

create_directories(output_base_dir)


Created directory structure under /content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_SPLIT


**Reasoning**:
I will iterate through the shuffled train, validation, and test file pairs and move the image files to their respective 'images' subdirectories and annotation files to their 'labels' subdirectories within the `output_base_dir` for each split.



In [9]:
import shutil
import os

# 9. Iterate through the shuffled file pairs and move them to their respective directories
def move_files(file_pairs, split_name, base_output_dir):
    split_images_dir = os.path.join(base_output_dir, split_name, "images")
    split_labels_dir = os.path.join(base_output_dir, split_name, "labels")

    print(f"Moving {len(file_pairs)} files to {split_name}...")
    for img_path, ann_path in file_pairs:
        # Move image file
        shutil.move(img_path, os.path.join(split_images_dir, os.path.basename(img_path)))
        # Move annotation file
        shutil.move(ann_path, os.path.join(split_labels_dir, os.path.basename(ann_path)))
    print(f"Finished moving files for {split_name}.")

# Move files for each split
move_files(train_pairs, "train", output_base_dir)
move_files(val_pairs, "val", output_base_dir)
move_files(test_pairs, "test", output_base_dir)

print("Data splitting and moving complete!")

Moving 2411 files to train...
Finished moving files for train.
Moving 689 files to val...
Finished moving files for val.
Moving 345 files to test...
Finished moving files for test.
Data splitting and moving complete!


## Generate YOLO YAML Configuration

### Subtask:
Create a `data.yaml` file in the correct YOLO format. This file will specify the paths to the training, validation, and testing image directories, the number of classes, and the class names.


**Reasoning**:
I will define the path to the 'classes.txt' file, which is located in the original input data directory, as it contains the class names required for the YOLO YAML configuration.



In [10]:
import os

# 1. Define the path to the 'classes.txt' file
classes_file_path = os.path.join(input_data_dir, 'classes.txt')

print(f"Classes file path: {classes_file_path}")

Classes file path: /content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_RAW/Perpaduan_Warna_Campuran/classes.txt


**Reasoning**:
I will read the class names from the 'classes.txt' file, storing each class name as an element in a list, which is essential for the YOLO configuration.



In [11]:
import os

# 2. Read the class names from the 'classes.txt' file
class_names = []
with open(classes_file_path, 'r') as f:
    for line in f:
        class_name = line.strip()
        if class_name:
            class_names.append(class_name)

num_classes = len(class_names)

print(f"Found {num_classes} classes: {class_names}")

Found 3 classes: ['Merah', 'Kuning', 'Hijau']


**Reasoning**:
I will define the paths for the training, validation, and testing image directories relative to the `output_base_dir`, which will be used in the `data.yaml` file.



In [12]:
import os

# 3. Define the paths for the training, validation, and testing image directories
#    relative to the output_base_dir for the YOLO YAML configuration.
train_img_path = os.path.join(output_base_dir, 'train', 'images')
val_img_path = os.path.join(output_base_dir, 'val', 'images')
test_img_path = os.path.join(output_base_dir, 'test', 'images')

print(f"Train images path: {train_img_path}")
print(f"Validation images path: {val_img_path}")
print(f"Test images path: {test_img_path}")

Train images path: /content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_SPLIT/train/images
Validation images path: /content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_SPLIT/val/images
Test images path: /content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_SPLIT/test/images


**Reasoning**:
I will construct the content of the `data.yaml` file as a string, incorporating the previously defined paths for training, validation, and testing image directories, the number of classes, and the class names, formatted correctly for YOLO.



In [13]:
import yaml

# 4. Construct the content of the data.yaml file as a string
data_yaml_content = {
    'train': train_img_path,
    'val': val_img_path,
    'test': test_img_path,  # Optional, but good practice to include
    'nc': num_classes,
    'names': class_names
}

yaml_string = yaml.dump(data_yaml_content, default_flow_style=False, sort_keys=False)

print("Generated data.yaml content:")
print(yaml_string)

Generated data.yaml content:
train: /content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_SPLIT/train/images
val: /content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_SPLIT/val/images
test: /content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_SPLIT/test/images
nc: 3
names:
- Merah
- Kuning
- Hijau



**Reasoning**:
I will create the `data.yaml` file in the `output_base_dir` and write the constructed YAML content string to it, completing the generation of the YOLO configuration file.



In [14]:
import os

# 5. Create the data.yaml file in the output_base_dir and write the constructed content to it.
data_yaml_path = os.path.join(output_base_dir, 'data.yaml')

with open(data_yaml_path, 'w') as f:
    f.write(yaml_string)

print(f"data.yaml successfully created at: {data_yaml_path}")

data.yaml successfully created at: /content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_SPLIT/data.yaml


## Final Task

### Subtask:
Confirm that the data split and YOLO configuration file have been successfully created and are ready for YOLOv11n training.


## Summary:

### Data Analysis Key Findings

*   The raw dataset located at `/content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_RAW/Perpaduan_Warna_Campuran` contained 6893 files, primarily consisting of `frame_XXXXX.png` image files and corresponding `frame_XXXXX.txt` annotation files, indicating a clear one-to-one pairing. A `classes.txt` file was also present.
*   A total of 3445 valid image-annotation pairs were identified for processing.
*   The dataset was successfully split into training, validation, and testing sets with a 70/20/10 ratio:
    *   Training set: 2411 samples
    *   Validation set: 689 samples
    *   Testing set: 345 samples
*   The split data was organized into a new directory structure under `/content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_SPLIT`, with separate `train`, `val`, and `test` subdirectories, each containing `images` and `labels` folders.
*   Three distinct classes were identified from `classes.txt`: 'Merah', 'Kuning', and 'Hijau'.
*   A `data.yaml` configuration file was successfully generated at `/content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_SPLIT/data.yaml`. It correctly specifies the paths for the training, validation, and testing image directories, sets `nc` (number of classes) to 3, and lists the class names.

### Insights or Next Steps

*   The dataset is now well-structured and properly configured for YOLOv11n training. The next logical step is to proceed with training a YOLO model using the prepared data and `data.yaml` configuration.
*   Consider performing data augmentation during training to further enhance model robustness, especially given the dataset size of 3445 samples, to mitigate overfitting and improve generalization.


# Task
Install the `Albumentations` library to prepare for image augmentation and handling bounding box transformations for object detection datasets.

## Install Albumentations Library

### Subtask:
Install the `Albumentations` library, which is suitable for performing image augmentations while also handling bounding box transformations for object detection datasets.


**Reasoning**:
I will install the `Albumentations` library and its dependencies using pip to prepare for image augmentations.



In [15]:
pip install -q -U albumentations opencv-python

**Reasoning**:
I will install the `Albumentations` library and its dependencies using pip to prepare for image augmentations.



In [16]:
pip install -q -U albumentations opencv-python

**Reasoning**:
I will install the `Albumentations` library and its dependencies using pip to prepare for image augmentations.



In [17]:
pip install -q -U albumentations opencv-python

**Reasoning**:
The Albumentations library and its dependencies have been installed in previous steps, so the next logical step is to confirm the successful installation by importing a module from it. This ensures that the installation was successful and the library is ready for use.



In [18]:
import albumentations as A
import cv2

print("Albumentations and OpenCV imported successfully.")

Albumentations and OpenCV imported successfully.


## Define Paths and Augmentation Parameters

### Subtask:
Set up the base directory for the split data and define paths to the training image and label directories. Define the percentage of training data to be augmented.


**Reasoning**:
I will define the base directory for the split data, construct paths to the training images and labels directories, and set the augmentation percentage, as required by the subtask.



In [19]:
import os

# 1. Set the base directory for the split dataset
output_base_dir = '/content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_SPLIT'

# 2. Define the path to the training images directory
train_images_dir = os.path.join(output_base_dir, 'train', 'images')

# 3. Define the path to the training labels directory
train_labels_dir = os.path.join(output_base_dir, 'train', 'labels')

# 4. Set the augmentation percentage
augmentation_percentage = 0.15

print(f"Output base directory: {output_base_dir}")
print(f"Training images directory: {train_images_dir}")
print(f"Training labels directory: {train_labels_dir}")
print(f"Augmentation percentage: {augmentation_percentage * 100}%")

Output base directory: /content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_SPLIT
Training images directory: /content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_SPLIT/train/images
Training labels directory: /content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_SPLIT/train/labels
Augmentation percentage: 15.0%


## Collect Training Image-Annotation Pairs

### Subtask:
Create a list of tuples, where each tuple contains the full path to an image file and its corresponding annotation file from the training set.


**Reasoning**:
I will get a list of all image files in the `train_images_dir` to prepare for iterating through them and finding their corresponding annotation files. This is the first step in collecting the training image-annotation pairs as per the instructions.



In [20]:
import os

# 1. Get a list of all image files in the train_images_dir
train_image_files = [f for f in os.listdir(train_images_dir) if f.endswith(('.png', '.jpg', '.jpeg'))]

# 2. Initialize an empty list to store the full paths of the image and its corresponding annotation file
train_image_annotation_pairs = []

print(f"Found {len(train_image_files)} image files in the training directory.")
print("First 5 training image files:")
for i, file_name in enumerate(train_image_files):
    if i >= 5: break
    print(f"  {file_name}")

Found 2411 image files in the training directory.
First 5 training image files:
  frame_00000.png
  frame_00001.png
  frame_00002.png
  frame_00005.png
  frame_00007.png


**Reasoning**:
I will iterate through each identified image file, construct its full path and the expected path for its corresponding annotation file. Then, I will verify the existence of both files and add their full paths as a tuple to the `train_image_annotation_pairs` list, and finally print a summary of the collected pairs.



In [21]:
import os

# 3. Iterate through each image file in the train_images_dir
for img_file_name in train_image_files:
    # 4. Construct its full path
    image_full_path = os.path.join(train_images_dir, img_file_name)

    # Get the base name without extension (e.g., 'frame_00000' from 'frame_00000.png')
    base_name = os.path.splitext(img_file_name)[0]

    # 5. Construct the expected full path for its corresponding annotation file
    annotation_file_name = base_name + '.txt'
    annotation_full_path = os.path.join(train_labels_dir, annotation_file_name)

    # 6. Check if both the image file and its corresponding annotation file exist
    if os.path.exists(image_full_path) and os.path.exists(annotation_full_path):
        # 7. If both exist, add a tuple (image_full_path, annotation_full_path) to the train_image_annotation_pairs list
        train_image_annotation_pairs.append((image_full_path, annotation_full_path))

# 8. Print the number of collected pairs and a few examples to verify
print(f"Found {len(train_image_annotation_pairs)} valid image-annotation pairs in the training set.")
print("First 5 training image-annotation pairs:")
for i, (img_path, ann_path) in enumerate(train_image_annotation_pairs):
    if i >= 5: break
    print(f"  Image: {os.path.basename(img_path)}, Annotation: {os.path.basename(ann_path)}")

Found 2411 valid image-annotation pairs in the training set.
First 5 training image-annotation pairs:
  Image: frame_00000.png, Annotation: frame_00000.txt
  Image: frame_00001.png, Annotation: frame_00001.txt
  Image: frame_00002.png, Annotation: frame_00002.txt
  Image: frame_00005.png, Annotation: frame_00005.txt
  Image: frame_00007.png, Annotation: frame_00007.txt


## Select Random Subset for Augmentation

### Subtask:
Randomly select the specified percentage (15%) of the training image-annotation pairs to undergo augmentation. This ensures a diverse subset of data is chosen for transformation.


**Reasoning**:
I will calculate the number of samples for augmentation, convert it to an integer, and then use `random.sample()` to select the specified percentage of `train_image_annotation_pairs`.



In [22]:
import random

# 1. Calculate the number of samples to augment
num_samples_to_augment = len(train_image_annotation_pairs) * augmentation_percentage

# 2. Convert the calculated number to an integer
num_samples_to_augment = int(num_samples_to_augment)

# 3. Use random.sample() to select a random subset of the train_image_annotation_pairs
samples_for_augmentation = random.sample(train_image_annotation_pairs, num_samples_to_augment)

print(f"Total training pairs: {len(train_image_annotation_pairs)}")
print(f"Number of samples to augment ({augmentation_percentage * 100}%): {num_samples_to_augment}")
print(f"Selected {len(samples_for_augmentation)} samples for augmentation.")
print("First 3 samples selected for augmentation:")
for i, (img_path, ann_path) in enumerate(samples_for_augmentation):
    if i >= 3: break
    print(f"  Image: {os.path.basename(img_path)}, Annotation: {os.path.basename(ann_path)}")

Total training pairs: 2411
Number of samples to augment (15.0%): 361
Selected 361 samples for augmentation.
First 3 samples selected for augmentation:
  Image: frame_Hijau_00309.png, Annotation: frame_Hijau_00309.txt
  Image: frame_Hijau_00702.png, Annotation: frame_Hijau_00702.txt
  Image: frame_Kuning_00454.png, Annotation: frame_Kuning_00454.txt


## Apply Augmentations and Save New Data

### Subtask:
Define a set of augmentation transforms (e.g., horizontal flip, small rotations, brightness/contrast changes) using Albumentations. Iterate through the selected subset of data, apply these transforms, and save the newly generated images and their corresponding YOLO format annotation files back into the training directories, ensuring unique filenames.


**Reasoning**:
I will define the augmentation pipeline using `albumentations.Compose` with specified transforms and `bbox_params` for YOLO format. This will prepare the augmentation object to be applied to the image and bounding box data.



In [23]:
import albumentations as A
import cv2

# 1. Define an augmentation pipeline using albumentations.Compose
transform = A.Compose([
    A.HorizontalFlip(p=0.5),
    A.RandomBrightnessContrast(p=0.2),
    A.ShiftScaleRotate(shift_limit=0.0625, scale_limit=0.1, rotate_limit=15, p=0.5),
    A.HueSaturationValue(hue_shift_limit=20, sat_shift_limit=30, val_shift_limit=20, p=0.2),
], bbox_params=A.BboxParams(format='yolo', label_fields=['class_ids']))

print("Augmentation pipeline defined.")

Augmentation pipeline defined.


/usr/local/lib/python3.12/dist-packages/albumentations/core/validation.py:114: UserWarning: ShiftScaleRotate is a special case of Affine transform. Please use Affine transform instead.
  original_init(self, **validated_kwargs)


**Reasoning**:
I will proceed with iterating through the selected samples, applying the defined augmentation pipeline, and saving the augmented images and their corresponding YOLO annotation files to the training directories, ensuring unique filenames as per the instructions.



In [24]:
import cv2
import os
import numpy as np

augmented_count = 0

for i, (image_path, annotation_path) in enumerate(samples_for_augmentation):
    # 4. Read the image using cv2.imread()
    image = cv2.imread(image_path)
    if image is None:
        print(f"Warning: Could not read image {image_path}. Skipping.")
        continue

    # 5. Read the annotation file and parse bounding boxes
    bboxes = []
    class_ids = []
    try:
        with open(annotation_path, 'r') as f:
            for line in f:
                parts = line.strip().split()
                if len(parts) == 5:
                    class_id = int(parts[0])
                    x_center, y_center, width, height = map(float, parts[1:])
                    bboxes.append([x_center, y_center, width, height])
                    class_ids.append(class_id)
    except FileNotFoundError:
        print(f"Warning: Annotation file {annotation_path} not found. Skipping.")
        continue
    except Exception as e:
        print(f"Warning: Error reading annotation file {annotation_path}: {e}. Skipping.")
        continue

    # 6. Apply the defined augmentation pipeline
    # Albumentations expects class_ids as a list for label_fields
    augmented = transform(image=image, bboxes=bboxes, class_ids=class_ids)
    augmented_image = augmented['image']
    augmented_bboxes = augmented['bboxes']
    augmented_class_ids = augmented['class_ids']

    # 7. Generate unique filenames for the augmented image and its corresponding annotation file
    original_base_name = os.path.splitext(os.path.basename(image_path))[0]
    new_image_name = f"{original_base_name}_aug_{augmented_count}.png"
    new_annotation_name = f"{original_base_name}_aug_{augmented_count}.txt"

    new_image_path = os.path.join(train_images_dir, new_image_name)
    new_annotation_path = os.path.join(train_labels_dir, new_annotation_name)

    # 8. Save the augmented image
    cv2.imwrite(new_image_path, augmented_image)

    # 9. Save the augmented bounding boxes and class IDs to a new annotation file
    with open(new_annotation_path, 'w') as f:
        for bbox, class_id in zip(augmented_bboxes, augmented_class_ids):
            # YOLO format: class_id x_center y_center width height
            f.write(f"{class_id} {bbox[0]} {bbox[1]} {bbox[2]} {bbox[3]}\n")

    # 10. Increment the augmented_count
    augmented_count += 1

# 11. After the loop, print the total number of new augmented samples created.
print(f"Successfully created {augmented_count} new augmented samples.")

Successfully created 361 new augmented samples.


## Verify Augmented Data Count

### Subtask:
Count the total number of image and annotation files in the training directories after augmentation to confirm that new data has been added and verify the increased dataset size.


**Reasoning**:
I will get a list of all image files in the `train_images_dir` after augmentation, count them, and print the count to confirm the increased dataset size.



In [25]:
import os

# 1. Get a list of all image files in the train_images_dir after augmentation
augmented_train_image_files = [f for f in os.listdir(train_images_dir) if f.endswith(('.png', '.jpg', '.jpeg'))]
augmented_image_count = len(augmented_train_image_files)

print(f"Total image files in training directory after augmentation: {augmented_image_count}")

Total image files in training directory after augmentation: 2772


**Reasoning**:
I will now get a list of all annotation files in the `train_labels_dir` after augmentation, count them, and print the count to address the second instruction of the subtask.



In [26]:
import os

# 2. Get a list of all annotation files in the train_labels_dir after augmentation
augmented_train_label_files = [f for f in os.listdir(train_labels_dir) if f.endswith('.txt')]
augmented_label_count = len(augmented_train_label_files)

print(f"Total annotation files in training directory after augmentation: {augmented_label_count}")

Total annotation files in training directory after augmentation: 2772


**Reasoning**:
I will now compare the augmented counts with the original training sample count to verify that new data has been successfully added, which is the final instruction of the subtask.



In [27]:
print(f"Original training samples (image-annotation pairs): {train_count}")
print(f"Total image files after augmentation: {augmented_image_count}")
print(f"Total annotation files after augmentation: {augmented_label_count}")

if augmented_image_count > train_count and augmented_label_count > train_count:
    print("Confirmation: Augmented data has been successfully added to the training set.")
else:
    print("Warning: Augmented data might not have been added as expected.")

Original training samples (image-annotation pairs): 2411
Total image files after augmentation: 2772
Total annotation files after augmentation: 2772
Confirmation: Augmented data has been successfully added to the training set.


## Final Task

### Subtask:
Summarize the data augmentation process, including the number of augmented samples created and confirm the updated state of the training dataset, ready for YOLOv11n training.


## Summary:

### Data Analysis Key Findings
*   The `Albumentations` library was successfully installed and verified for image augmentation, including bounding box transformations.
*   The initial training dataset contained 2411 image-annotation pairs.
*   A total of 361 samples (15% of the original training data) were randomly selected for augmentation.
*   An augmentation pipeline was applied, including horizontal flip, random brightness/contrast, shift/scale/rotate, and hue/saturation/value transformations.
*   **361 new augmented image and annotation pairs were successfully generated and saved** to the training directories.
*   The final training dataset size increased from 2411 to **2772 image-annotation pairs**, confirming the successful addition of augmented data.

### Insights or Next Steps
*   The expanded training dataset, now containing 2772 samples, is ready for training a YOLOv11n object detection model. This augmentation should contribute to improved model robustness and generalization.


# Task
Install the `ultralytics` library to enable access to YOLOv11 models and their training functionalities.

## Install Ultralytics Library

### Subtask:
Install the `ultralytics` library, which provides access to YOLOv11 models and their training functionalities. This is a prerequisite for training the model.


**Reasoning**:
I will install the `ultralytics` library using pip, including the `-q` and `-U` flags for a quiet and upgrade installation as per the instructions.



In [28]:
pip install -q -U ultralytics

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 25.6 MB/s eta 0:00:00


**Reasoning**:
Now that the `ultralytics` library has been installed, I will confirm the successful installation by importing the `YOLO` class from the `ultralytics` module. This will ensure that the library is correctly set up and ready for use.



In [29]:
from ultralytics import YOLO

print("Ultralytics YOLO imported successfully.")

Creating new Ultralytics Settings v0.0.6 file ✅ 
View Ultralytics Settings with 'yolo settings' or at '/root/.config/Ultralytics/settings.json'
Update Settings with 'yolo settings key=value', i.e. 'yolo settings runs_dir=path/to/dir'. For help see https://docs.ultralytics.com/quickstart/#ultralytics-settings.
Ultralytics YOLO imported successfully.


## Load YOLOv11n Model

### Subtask:
Import the `YOLO` class from `ultralytics` and load a pre-trained YOLOv11n model. This model will serve as the base for training on your custom dataset.


**Reasoning**:
I will load the pre-trained YOLOv11n model using the YOLO class and store it in a variable named 'model' as per the instructions.



In [30]:
from ultralytics import YOLO

# 1. Load a pre-trained YOLOv11n model
# 2. Store the loaded model in a variable named 'model'
model = YOLO('yolov11n.pt')

print("YOLOv11n model loaded successfully.")

FileNotFoundError: [Errno 2] No such file or directory: 'yolov11n.pt'

**Reasoning**:
The previous code failed because it was looking for 'yolov11n.pt' locally, but the ultralytics library can download the model weights automatically if just the model name is provided. I will correct the model loading call to trigger the download.



In [31]:
from ultralytics import YOLO

# 1. Load a pre-trained YOLOv11n model
# 2. Store the loaded model in a variable named 'model'
model = YOLO('yolov11n')

print("YOLOv11n model loaded successfully.")

FileNotFoundError: 'yolov11n' does not exist

**Reasoning**:
The ultralytics library expects the model name with a '.pt' extension for pre-trained models to trigger an automatic download if the file is not found locally. I will update the model name to `yolov8n.pt`, which is a commonly available nano model in the Ultralytics library, assuming `yolov11n` was a placeholder or a typo for the latest nano version.



In [35]:
from ultralytics import YOLO

# 1. Load a pre-trained YOLOv8n model
# 2. Store the loaded model in a variable named 'model'
model = YOLO('yolo11n.pt')

print("YOLOv11n model loaded successfully.")

YOLOv11n model loaded successfully.


## Configure and Train YOLOv8n Model

### Subtask:
Define the training parameters such as the number of epochs, image size, and batch size. Then, initiate the training process using the loaded YOLOv8n model and the provided `data.yaml` configuration file.


**Reasoning**:
I will define the training parameters and initiate the YOLOv8n model training using the `train()` method, passing the `data.yaml` configuration and the specified parameters.



In [36]:
import os

# 1. Define the training parameters
# Assuming data_yaml_path and model are already defined from previous steps

# Get the path to data.yaml from previous steps
# output_base_dir = '/content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_SPLIT'
# data_yaml_path = os.path.join(output_base_dir, 'data.yaml')

# Define training arguments
training_args = {
    'data': data_yaml_path,  # Path to the data.yaml file
    'epochs': 50,          # Number of training epochs
    'imgsz': 640,           # Input image size (e.g., 640 for 640x640 images)
    'batch': 16,            # Batch size
    'name': 'yolov8n_custom_color_detection', # Name for the training run
    'patience': 10,          # Number of epochs to wait without improvement before stopping training
    'lr0': 0.001,
    'optimizer': 'AdamW'
}

# 2. Initiate the training process
results = model.train(**training_args)

print("YOLOv11n training initiated with custom dataset and parameters.")

Ultralytics 8.3.239 🚀 Python-3.12.12 torch-2.9.0+cu126 CUDA:0 (Tesla T4, 15095MiB)
engine/trainer: agnostic_nms=False, amp=True, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/content/drive/MyDrive/Percobaan_Trainning_Warna/DATA_SPLIT/data.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, epochs=50, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.001, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolo11n.pt, momentum=0.937, mosaic=1.0, multi_scale=False, name=yolov8n_custom_color_detection2, nbs=64, nms=False, opset=None, optimize=False, optimizer=AdamW,

# Task
Define the source directory of the YOLOv8n training results as `/content/runs/detect/yolov8n_custom_color_detection2` and the destination directory for copying these results to Google Drive as `/content/drive/MyDrive/Percobaan_Trainning_Warna/YOLOv8n_Training_Results`.

## Define Source and Destination Paths

### Subtask:
Define the source directory (output of YOLOv8n training) and the destination directory (Google Drive folder) for copying.


**Reasoning**:
I will define the source and destination directory paths as instructed, which are essential for identifying the training output and the target location in Google Drive.



In [37]:
import os

# 1. Define the source directory (output of YOLOv8n training)
source_dir = '/content/runs/detect/yolov8n_custom_color_detection2'

# 2. Define the destination directory (Google Drive folder)
destination_dir = '/content/drive/MyDrive/Percobaan_Trainning_Warna/YOLOv8n_Training_Results'

print(f"Source directory: {source_dir}")
print(f"Destination directory: {destination_dir}")

Source directory: /content/runs/detect/yolov8n_custom_color_detection2
Destination directory: /content/drive/MyDrive/Percobaan_Trainning_Warna/YOLOv8n_Training_Results


## Copy Training Results

### Subtask:
Copy the entire contents of the source directory to the destination directory. This includes trained weights, plots, and other training artifacts.


**Reasoning**:
I will import the `shutil` module, then check if the destination directory already exists. If it does, I will remove it to ensure a clean copy. After that, I will copy the entire contents from the source directory to the destination directory using `shutil.copytree`.



In [38]:
import shutil
import os

# Ensure the destination directory exists, if not, create it. If it exists, remove and recreate to ensure a fresh copy.
if os.path.exists(destination_dir):
    print(f"Removing existing destination directory: {destination_dir}")
    shutil.rmtree(destination_dir)

# Copy the entire contents of the source directory to the destination directory
shutil.copytree(source_dir, destination_dir)

print(f"Successfully copied contents from '{source_dir}' to '{destination_dir}'.")

Successfully copied contents from '/content/runs/detect/yolov8n_custom_color_detection2' to '/content/drive/MyDrive/Percobaan_Trainning_Warna/YOLOv8n_Training_Results'.


## Verify Copied Files (Optional)

### Subtask:
List the contents of the destination directory to ensure that the files have been copied successfully.


**Reasoning**:
I will list the contents of the `destination_dir` using `os.listdir()` and print the total number of items, along with the first 10 items, to verify the successful copy operation as per the instructions.



In [39]:
import os

# 1. Use os.listdir() to get a list of all files and subdirectories within the destination_dir.
contents_of_destination = os.listdir(destination_dir)

# 2. Print the total number of items found in the directory.
print(f"Total items found in '{destination_dir}': {len(contents_of_destination)}")

# 3. Print the first 10 items from the list to get an overview of the copied contents.
print("First 10 items in the destination directory:")
for i, item in enumerate(contents_of_destination):
    if i >= 10:
        break
    print(f"  {item}")

Total items found in '/content/drive/MyDrive/Percobaan_Trainning_Warna/YOLOv8n_Training_Results': 23
First 10 items in the destination directory:
  BoxF1_curve.png
  BoxP_curve.png
  results.png
  train_batch6961.jpg
  confusion_matrix.png
  train_batch6960.jpg
  BoxPR_curve.png
  val_batch1_pred.jpg
  train_batch2.jpg
  results.csv


## Final Task

### Subtask:
Summarize the completion of copying the YOLOv8n training results to the specified Google Drive folder.


## Summary:

### Data Analysis Key Findings
*   The YOLOv8n training results were successfully copied from the local Colab environment directory `/content/runs/detect/yolov8n_custom_color_detection2` to the Google Drive folder `/content/drive/MyDrive/Percobaan_Trainning_Warna/YOLOv8n_Training_Results`.
*   The copying process ensured data integrity by first removing any existing destination directory before performing a fresh copy of the entire contents, including trained weights, plots, and other training artifacts.
*   Verification confirmed that a total of 23 items, including essential training output files like `BoxF1_curve.png`, `results.png`, `confusion_matrix.png`, and `results.csv`, were successfully transferred to the specified Google Drive location.

### Insights or Next Steps
*   The successful transfer of YOLOv8n training results to Google Drive ensures data persistence and accessibility, allowing for long-term storage, sharing, and further analysis outside the Colab environment.
*   The availability of all training artifacts in Google Drive facilitates easy review of model performance, comparison with other training runs, and potential deployment of the trained model.
